src/Build.UnitTests/Instance/ProjectInstance_Internal_Tests.cs

```diff
@@ -480,8 +481,8 @@ public void CloneTaskRegistry()
             ProjectInstance first = GetSampleProjectInstance();
             ProjectInstance second = first.DeepCopy();
 
-            // Task registry object should be immutable
```

Comment:
Was this assertion always wrong? Have we added mutability where we shouldn't have, or just "in a way that requires we change this"?

Reply:
Unfortunately this is older then GH history so I can only provide guessing - and my guess is the ProjectInstance and owned object graph (including TaskRegistry) was always meant to be instantiated once and then used only sequantially. So immutability and deepclonning was never strongly followed - as it was mostly ok in majority of cases (a specific set of VS actions and conditions needs to be made - e.g. in-proc build or cancel and rebuild - to hit the unwanted sharing of TaskRegistry/Toolset).
TaskRegistry looks _almost_ immutable, with exception of the internal matching caches, lazy initialized state of held TaskFactoryWrapper instances (the reason for the 3rd state corruption perf issue), plus something else (I forgot now) - so the path of DeepCopying turned out to be easier and safer than attempting to make TaskRegistry and Toolset truly immutable

Reply:
tl;dr; the added mutability shouldn't be added without adding deep cloning IMHO